[RFC] Enable -Wconversion: if_cscope.c

glindstedt opened the pull request:
Enable -Wconversion for if_cscope.c as per #567. Mostly includes some casts and type-changes from `int` to `size_t`, where I've added asserts to make sure the value is non-negative.

I chose to keep the global variable `csinfo_size` as an `int` since it's used extensively within loops and the amount of casts are relatively small.

In the function `cs_add_common()` I've left the `len` variable as an `int` for now since it is used as such in a call to `modify_fname()` on line 469. `modify_fname()` is defined in `eval.c` and only invoked 4 times in the source. The parameter is a length and is not used in any loops, so I would suggest changing the parameter to `size_t` instead. If that sounds like a good idea I can take care of it, it doesn't seem like anyone is messing with eval.c yet.

glindstedt committed:
Enable -Wconversion: if_cscope.c

justinmk commented on src/nvim/if_cscope.c:
```diff
@@ -1772,11 +1777,14 @@ static int cs_read_prompt(int i)
           /* append character to the message */
-          buf[bufpos++] = ch;
+          assert(ch >= 0);
+          assert(ch <= CHAR_MAX);
```
Can be in one assert with `&&`.

glindste replied:
Ok, will merge them!

Pyrohh replied:
Wouldn't that make the assert less meaningful? Take this example:

```
#include <assert.h>

int main(int argc, char **argv)
{
    int i = 0;
    assert(i >= 1 && i <= -1 && i != 5 && i != 61);
    return 0;
}
```

It aborts with this message:
```
a.out: lel.c:6: main: Assertion `i >= 1 && i <= -1 && i != 5 && i != 61' failed.
```

If we had only one clause per assert statement, then the abort message would be more meaningful.

I'm not sure how many implementations of `assert.h` do this, but it can be implemented as a macro, so there's no function call overhead (although that argument might be redundant as it would probably be inlined anyways).

justinmk replied:
If I recall, we do this elsewhere. And the crashdump will have the stack values, so the granularity isn't needed, I think.

Pyrohh replied:
> And the core dump will have the stack values, so the granularity isn't needed, I think.

While much simpler (in some places) than Vim, it does seem like Nvim is too complicated to get much meaning out of an assert. Still, I don't see any reason to use `&&`, besides the line savings (which IMO doesn't matter because they're harder to read).

justinmk replied:
asserts are extremely valuable, at least for the core dump they generate. assert() is already a fairly ugly compromise, so tidying it up where we can seems worth it to me.

Pyrohh replied:
> asserts are extremely valuable, at least for the core dump they generate

I didn't mean to imply that asserts as a whole aren't meaningful (with core-dumps and whatnot), I meant that the statements printed by asserts aren't necessarily useful in themselves, as shown by the example above. 

> assert() is already a fairly ugly compromise, so tidying it up where we can seems worth it to me.

I'd argue that we should prefer practicality over style in this situation, but I'm the kind of person who thinks the comma operator is evil so I'm possible too biased :)

justinmk replied:
I agree but... the assert message doesn't *have* much practical value. That's why there the extra noise in the source is not worth it.

Pyrohh replied:
:+1:

elmart replied:
To me, one single assert is more conceptually clear, and reads better.
There are complicated asserts with multiple clauses. Augmenting code pollution by multiplying lines doesn't look like a good idea to me.
If you get a bug report for a failing assert, it's easy enough to split it then if you want, while debugging.
Also, note that many of the asserts will be simplified when dedicated macros to check for bounds/overflow are introduced.

deleisha replied:
I am very new to neovim. So my comments might not be best practice for neovim.
Does neovim accepts following assert style? 
I would rather go with single assert like below
assert( (ch >= CHAR_MIN && ch <= CHAR_MAX) && "ch is not in range of char");

elmart replied:
It is planned to replace those range-checking conditions by dedicated macros at some point.
Then, it would read like this: `assert(IN_RANGE(INT, ch));`, which wouldn't need further explanation.
For asserts requiring explanation, comments are preferred.

elmart commented on src/nvim/if_cscope.c:
```diff
@@ -775,17 +778,17 @@ static int cs_create_connection(int i)
 
     /* alloc space to hold the cscope command */
-    len = (int)(strlen(prog) + strlen(csinfo[i].fname) + 32);
+    len = (strlen(prog) + strlen(csinfo[i].fname) + 32);
```
Parens not needed anymore.

glindste replied:
Thanks, missed that one.

elmart commented:
I'd try to refactor `csinfo_size` into `size_t`, as well as `totmatches`. Doesn't seem complicated, and it would be restricted to current file.

> In the function cs_add_common() I've left the len variable as an int for now since it is used as such in a call to modify_fname() on line 469. modify_fname() is defined in eval.c and only invoked 4 times in the source. The parameter is a length and is not used in any loops, so I would suggest changing the parameter to size_t instead. If that sounds like a good idea I can take care of it, it doesn't seem like anyone is messing with eval.c yet.

I'm currently doing it (refactoring `modify_fname` param into `size_t`), as a byproduct of enabling -Wconversion in normal.c. So, don't do it this time (to avoid repetition of work). But, in general, I'd avice to try to cascade these easy changes as far as possible, while it doesn't get too much complicated to review.

glindste commented:
Ok, I'll try and refactor them. I'll leave `modify_fname` to you then, thanks for the heads up!

elmart commented:
> Ok, I'll try and refactor them

:+1: I tried doing it locally (not completely, just to have a feeling about it) before advisingRefactoring involved loop counters to `size_t` too makes everything smooth and poses no problems.

glindstedt committed:
Enable -Wconversion: ifcscope refactor csinfo_size

Refactor csinfo_size, totmatches and other related variables and
functions to use size_t instead of int.

glindste commented:
I've refactored `csinfo_size`, `totmatches` as well as a lot of other variables and functions that were affected. In most cases it went smooth, although I had to restructure `cs_kill` a bit. I'd be happy to get some comments on how to improve it.

The `nummatches` array is still an `int` array as of now which leads to a cast. This is due to the values in the array being set using `cs_cnt_matches` which can return a negative value. I could look in to refactoring that as well later.

I've tried some basic stuff to test it like adding/deleting databases and simple searches and it seems to work fine.

elmart commented on src/nvim/if_cscope.c:
```diff
@@ -127,7 +127,7 @@ char_u *get_cscope_name(expand_T *xp, int idx)
         continue;
       if (current_idx++ == idx) {
-        vim_snprintf(connection, sizeof(connection), "%d", i);
+        vim_snprintf(connection, sizeof(connection), "%lu", i);
```
This should "%zu" for C99.

elmart replied:
There are more like this.
Sorry, I must go now. Will continue reviewing later.

glindste replied:
Yes, I used "%zu" where the regular `printf` functions were used, but on line 3042 in `message.c` there is a comment about what `snprintf` supports, and it has no mention of "%zu". I thought that "%lu" would be the next best thing. If it in fact does support "%zu" then I'll change it, I did not delve in to the source to check, but I could do that to verify.

Pyrohh commented on src/nvim/if_cscope.c:
```diff
@@ -1454,7 +1477,7 @@ static char *cs_manage_matches(char **matches, char **contexts, int totmatches,
  * parse cscope output
  */
-static char *cs_parse_results(int cnumber, char *buf, int bufsize, char **context, char **linenumber, char **search)
+static char *cs_parse_results(size_t cnumber, char *buf, int bufsize, char **context, char **linenumber, char **search)
```
Could you break this line please? Thankfully, lines of this length are becoming a rarity.

glindste replied:
:+1:

Pyrohh commented on src/nvim/if_cscope.c:
```diff
@@ -1509,7 +1532,8 @@ static char *cs_parse_results(int cnumber, char *buf, int bufsize, char **contex
 {
-  int i, j;
+  size_t i;
+  int j;
```
These could probably be moved down to their respective initializations.

glindste replied:
In this case I think it's a good idea actually, and I can go through the other cases. However sometimes, for example when a loop counter is reused in several loops which are not nested, I would argue that declaring them at the start is more readable.

Pyrohh replied:
> However sometimes, for example when a loop counter is reused in several loops which are not nested, I would argue that declaring them at the start is more readable.

I suppose it increases readability from the perspective of the lines being shorter, but IMO it's more obvious that a for loop counter isn't being used anywhere else if it's declared inline.

glindste replied:
My first thought was that you could see all the variables used in the function by just looking at the top of the function. But I think I've changed my mind, and I agree that it is more readable if defined inline in the loop declaration in this case. Because declaring the variable inside the loop declaration also limits the variable's scope, so that you know that it doesn't remain relevant beyond the loop.

glindste commented:
I'll be away until sunday so won't be able to respond until then.

glindstedt committed:
Enable -Wconversion: ifcscope tidy up and refactor

Break some long function definitions, move some variable declarations
down to initializations, fix formatting in cs_pathcomponents(). Change
some variables from int to size_t in cs_read_prompt().

Pyrohh commented on src/nvim/if_cscope.c:
```diff
@@ -933,12 +934,14 @@ static int cs_find(exarg_T *eap)
-static int cs_find_common(char *opt, char *pat, int forceit, int verbose, int use_ll, char_u *cmdline)
+static int cs_find_common(
+  char *opt, char *pat, int forceit,
+  int verbose, int use_ll, char_u *cmdline
```
Thanks for breaking these, but this isn't the style that we use IIRC, as shown here:

http://neovim.org/develop/style-guide.xml#Function_Declarations_and_Definitions

glindste replied:
Ah, thanks. I'll amend the commit. I'll read through the style guide.

glindstedt committed:
Enable -Wconversion: ifcscope minor formatting fix

Make sure the changes made to if_cscope.c in order to enable
-Wconversion adheres to the style guidelines.

elmart commented:
Do you plan to continue working on this?

glindste commented:
Yes, but I'm not sure where to continue at this moment. I was waiting for more comments at first, and recently I haven't had that much time, but I was planning to ask soon about what to do next. I could continue refactoring and trying make the code follow the style guidelines, but I felt like I was moving outside the scope of this pull request.

And regarding your earlier comment about "%zu", I looked at the code for `vim_snprintf` and couldn't see any sign of support for it. Do you think "%lu" is good enough for now?

Anyway I was thinking that since you were in the process of refactoring `modify_fname()` to take `size_t` it would be best to wait with merging this pull request until that has been merged and then rebase, so I can modify `cs_add_common()` to it's final state.

elmart commented:
@glindste 

Ok. I didn't want to put pressure on you. Just checking this PR wasn't stalled.

Regarding `%zu`, you're right. It seems it's not currently supported by `vim_snprintf`. If you don't mind, I'll give it a try to add support for that.

> Anyway I was thinking that since you were in the process of refactoring modify_fname() to take size_t it would be best to wait with merging this pull request until that has been merged and then rebase, so I can modify cs_add_common() to it's final state.

Ok. I'll come back when that and the `vim_snprintf` improvement is merged, then.

Thanks.

Pyrohh commented:
Given the above discussion, I'll set this to WIP. Feel free to change it back to RFC whenever you feel it's ready @glindst, or if you think I've made a mistake.

glindste commented:
@elmart Ok, no worries. Sounds good!

glindste commented:
@Pyrohh Yeah sounds like a good idea :+1:

glindste commented:
Rebased. Nice work @elmart, thanks!

elmart commented:
You're welcome. I'm sorry this took so long. I hadn't notified you because I'm still waiting for #2490 to be merged.

glindste commented:
@elmart don't worry about it, there's no rush :) I can see if I can help with reviewing #2490

elmart commented:
Ok, you can proceed now. Thanks!

glindstedt committed:
Enable -Wconversion: ifcscope use %zu for size_t

Use %zu when printing size_t with vim_snprintf()

glindste commented:
Sorry for the delay, rebased and added the last commit that uses `%zu` when printing with `vim_snprintf`.

Pyrohh commented:
Not sure what @elmart thinks, but could you squash the `%zu` changes into the other commits?

glindste commented:
@Pyrroh I assumed all commits would be squashed into one before merge so I didn't consider it, but I could squash them now if it makes reviewing easier.

Pyrohh commented:
It's fine, if it's going to be squashed then there's no need.

justinmk commented:
Restarted build since travis was upgraded.

justinmk commented on src/nvim/if_cscope.c:
```diff
@@ -1748,20 +1761,20 @@ static void cs_print_tags_priv(char **matches, char **cntxts, int num_matches)
   cs_emsg = _("E609: Cscope error: %s");
   /* compute maximum allowed len for Cscope error message */
-  maxlen = (int)(IOSIZE - strlen(cs_emsg));
+  assert(IOSIZE > strlen(cs_emsg));
```
Shouldn't this be `IOSIZE >= ...` ?

glindste replied:
Yes, good catch. Since I'll make a change anyway I'll make sure it's stored in a variable.

splinterofchaos commented on src/nvim/if_cscope.c:
```diff
@@ -1420,12 +1436,12 @@ static char *cs_manage_matches(char **matches, char **contexts, int totmatches,
-            xfree(cp[cnt]);
-        }
+      while (cnt > 0) {
+        cnt--;
```
I think `while (cnt--)` made more sense here. We don't need to worry about decrementing it below zero because we reset it after this loop.

glindste replied:
Yes, good point. I don't remember why I moved it out in the first place. I was probably worried about underflow but when it reaches 0 the loop will stop anyway.

splinterofchaos commented on src/nvim/if_cscope.c:
```diff
@@ -1615,19 +1632,15 @@ static void cs_fill_results(char *tagstr, int totmatches, int *nummatches_a, cha
   if (p_cspc == 0)
     return path;
 
   s = path + strlen(path) - 1;
```
Could you move the declaration of `s` here?

glindste replied:
:+1:

glindstedt committed:
Enable -Wconversion: ifcscope fix comparison bug

Fix comparison bug and a few minor style issues

glindste commented:
Rebased and applied suggestions.

justinmk commented on src/nvim/if_cscope.c:
```diff
@@ -1174,22 +1172,25 @@ static char *GetWin32Error(void)
+  size_t i;
+  bool empty_found;
 
-  i = -1;   /* can be set to the index of an empty item in csinfo */
```
QB build complains that `i` is used uninitialized in `cs_insert_filelist`. I'm guessing it is because of this deletion.

glindste replied:
Oh thanks! I'll fix that :+1:

glindstedt committed:
Enable -Wconversion: ifcscope fix warning

Fix warning of uninitialized variable in cs_insert_filelist.

justinmk commented:
Merged to https://github.com/neovim/neovim/pull/3068
